### ISSUE TITLE
Run the code and send ss

### ISSUE THREAD

**Issue by kkiritok** [kkiritok/Quantam-based-optimization#17]
# ======================
# IMPORTS
# ======================
import os
import tarfile
import pandas as pd
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader
from torchvision import transforms
from torchvision.models import resnet18
from PIL import Image
from sklearn.model_selection import train_test_split

# ======================
# DEVICE
# ======================
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

# ======================
# PATHS
# ======================
BASE_PATH = "/kaggle/input/datasets/zcyzhchyu/mini-imagenet"
IMAGE_TAR_PATH = os.path.join(BASE_PATH, "images.tar")
EXTRACT_PATH = "/kaggle/working/images"

# ======================
# EXTRACT IMAGES
# ======================
if not os.path.exists(EXTRACT_PATH):
    os.makedirs(EXTRACT_PATH, exist_ok=True)
    with tarfile.open(IMAGE_TAR_PATH) as tar:
        tar.extractall(EXTRACT_PATH)

print("Images extracted!")

# ======================
# LOAD CSV FILES
# ======================
train_df = pd.read_csv(os.path.join(BASE_PATH, "train.csv"))
val_df   = pd.read_csv(os.path.join(BASE_PATH, "val.csv"))
test_df  = pd.read_csv(os.path.join(BASE_PATH, "test.csv"))

# Combine all
full_df = pd.concat([train_df, val_df, test_df], ignore_index=True)

print("Total samples:", len(full_df))
print("Total classes:", full_df['label'].nunique())

# ======================
# LABEL ENCODING
# ======================
labels = sorted(full_df['label'].unique())
label2idx = {l: i for i, l in enumerate(labels)}
full_df['label'] = full_df['label'].map(label2idx)

# ======================
# TRAIN-TEST SPLIT
# ======================
train_df, test_df = train_test_split(
    full_df,
    test_size=0.2,
    stratify=full_df['label'],
    random_state=42
)

print("Train size:", len(train_df))
print("Test size :", len(test_df))

# ======================
# CHECK COLUMN NAME
# ======================
print("Columns:", train_df.columns)

# 👉 Change this if needed
IMAGE_COL = "filename"   # or "image_name" / "file"

# ======================
# CUSTOM DATASET
# ======================
class MiniImageNetDataset(Dataset):
    def __init__(self, df, image_dir, transform=None):
        self.df = df.reset_index(drop=True)
        self.image_dir = image_dir
        self.transform = transform

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        img_name = self.df.loc[idx, IMAGE_COL]
        label = self.df.loc[idx, 'label']

        img_path = os.path.join(self.image_dir, img_name)

        image = Image.open(img_path).convert("RGB")

        if self.transform:
            image = self.transform(image)

        return image, label

# ======================
# TRANSFORMS
# ======================
train_transform = transforms.Compose([
    transforms.RandomResizedCrop(224),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

test_transform = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225]
    )
])

# ======================
# DATASET + DATALOADER
# ======================
IMAGE_PATH = os.path.join(EXTRACT_PATH, "images")  # adjust if needed

train_dataset = MiniImageNetDataset(train_df, IMAGE_PATH, train_transform)
test_dataset  = MiniImageNetDataset(test_df, IMAGE_PATH, test_transform)

trainloader = DataLoader(
    train_dataset,
    batch_size=128,
    shuffle=True,
    num_workers=4,
    pin_memory=True
)

testloader = DataLoader(
    test_dataset,
    batch_size=128,
    shuffle=False,
    num_workers=4,
    pin_memory=True
)

num_classes = len(labels)
print("Classes:", num_classes)

# ======================
# MODEL
# ======================
model = resnet18(pretrained=True)

# Freeze ONLY layer4
for param in model.layer4.parameters():
    param.requires_grad = False

model.fc = nn.Linear(512, num_classes)
model = model.to(device)

criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.AdamW(
    filter(lambda p: p.requires_grad, model.parameters()),
    lr=3e-4,
    weight_decay=1e-4
)

# ======================
# ACCURACY FUNCTION
# ======================
def accuracy(model, loader):
    model.eval()
    correct = 0
    total = 0

    with torch.no_grad():
        for x, y in loader:
            x, y = x.to(device), y.to(device)

            logits = model(x)
            preds = logits.argmax(dim=1)

            correct += (preds == y).sum().item()
            total += y.size(0)

    return 100 * correct / total

# ======================
# TRAINING LOOP
# ======================
EPOCHS = 10

print("\n🚀 TRAINING STARTED (Mini-ImageNet)")
for epoch in range(EPOCHS):
    model.train()
    loss_sum = 0.0

    for x, y in trainloader:
        x, y = x.to(device), y.to(device)

        optimizer.zero_grad()

        logits = model(x)
        loss = criterion(logits, y)

        loss.backward()
        optimizer.step()

        loss_sum += loss.item()

    train_acc = accuracy(model, trainloader)
    test_acc = accuracy(model, testloader)

    print(
        f"Epoch {epoch+1}/{EPOCHS} | "
        f"Loss {loss_sum/len(trainloader):.4f} | "
        f"Train Acc {train_acc:.2f}% | "
        f"Test Acc {test_acc:.2f}%"
    )
